[JSC] Promise.resolve should be folded into fulfill

Constellation opened the pull request:
#### 9aa1cb79e9dbfd91ef0a1f603a8fb5594a6df833
<pre>
[JSC] Promise.resolve should be folded into fulfill
<a href="https://bugs.webkit.org/show_bug.cgi?id=300550">https://bugs.webkit.org/show_bug.cgi?id=300550</a>
<a href="https://rdar.apple.com/162424524">rdar://162424524</a>

Reviewed by Sosuke Suzuki.

`Promise.resolve(value)` specially handles &quot;thenable&quot; objects which
have &quot;then&quot; function. This is kind-of backward compatibility feature
so that native Promise can work together with non native Promise
implementations. However, to detect &quot;thenable&quot;, we need to check
&quot;then&quot; function existence for any input objects and this is costly
since (1) now almost everyone is using native Promise so &quot;thenable&quot;
custom promise is rare, and (2) we need to check &quot;then&quot; function for any
object argument which is just a resolution value for the promise.

In this patch, we add intrinsics for resolve operations and handle them
in DFG / FTL. And through DFG pipeline, we attempt to fold &quot;resolve&quot;
operation into &quot;fulfill&quot; operation which does not pipe the value as a
promise. We convert &quot;resolve&quot; to &quot;fulfill&quot; when,

1. Input is not an object. Then we do not need to look up &quot;then&quot;.
2. Input Structure is finite, it is only one and we cannot find &quot;then&quot;
   function. We insert watchpoint for this case to ensure this shape and
   convert &quot;resolve&quot; to &quot;fulfill&quot;.

Tests: JSTests/stress/promise-resolve-folded-own.js
       JSTests/stress/promise-resolve-folded-proto.js

* JSTests/stress/promise-resolve-folded-own.js: Added.
(test):
(object.then):
* JSTests/stress/promise-resolve-folded-proto.js: Added.
(test):
(Object.prototype.then):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
(JSC::DFG::ConstantFoldingPhase::foldConstants):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGGraph.cpp:
(JSC::DFG::Graph::tryEnsureAbsence):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):

Canonical link: <a href="https://commits.webkit.org/301376@main">https://commits.webkit.org/301376@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e9381744dd77c53dee1614e684d488645be032c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125765 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45427 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36178 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132630 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77645 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4421e96b-e55e-4894-a38f-dc5eee0a7c4b) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46110 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53986 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/95813 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63931 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/72c86cd8-bed2-4510-9b78-8282e29a56e4) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128713 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36865 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112466 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76305 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/70ac0851-fbb0-4bab-8c9a-088e54ccf384) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35768 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30651 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76100 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/117847 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106644 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30864 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135309 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Compiled WebKit") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/124272 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52552 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40307 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/104277 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53000 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108677 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/104007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49372 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27690 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49847 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19688 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52447 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58256 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/157285 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51797 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39372 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55145 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53491 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on JSTests/stress/promise-resolve-fold-own.js:
```diff
@@ -0,0 +1,17 @@
+}
+
+object.then = function () {
+    throw new Erorr("Hey");
```
`Erorr` -> Error` ?

Constellation replied:
Fixed.

sosukesuzuki commented on JSTests/stress/promise-resolve-fold.js:
```diff
@@ -0,0 +1,17 @@
+}
+
+Object.prototype.then = function () {
+    throw new Erorr("Hey");
```
ditto

Constellation replied:
Fixed.

sosukesuzuki reviewed (commented):
`fold` in the commit title should be `folded` ?

sosukesuzuki reviewed (approved):
r=me

Constellation committed:
[JSC] Promise.resolve should be folded into fulfill
https://bugs.webkit.org/show_bug.cgi?id=300550
rdar://162424524

Reviewed by Sosuke Suzuki.

`Promise.resolve(value)` specially handles "thenable" objects which
have "then" function. This is kind-of backward compatibility feature
so that native Promise can work together with non native Promise
implementations. However, to detect "thenable", we need to check
"then" function existence for any input objects and this is costly
since (1) now almost everyone is using native Promise so "thenable"
custom promise is rare, and (2) we need to check "then" function for any
object argument which is just a resolution value for the promise.

In this patch, we add intrinsics for resolve operations and handle them
in DFG / FTL. And through DFG pipeline, we attempt to fold "resolve"
operation into "fulfill" operation which does not pipe the value as a
promise. We convert "resolve" to "fulfill" when,

1. Input is not an object. Then we do not need to look up "then".
2. Input Structure is finite, it is only one and we cannot find "then"
   function. We insert watchpoint for this case to ensure this shape and
   convert "resolve" to "fulfill".

Tests: JSTests/stress/promise-resolve-folded-own.js
       JSTests/stress/promise-resolve-folded-proto.js

* JSTests/stress/promise-resolve-folded-own.js: Added.
(test):
(object.then):
* JSTests/stress/promise-resolve-folded-proto.js: Added.
(test):
(Object.prototype.then):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
(JSC::DFG::ConstantFoldingPhase::foldConstants):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGGraph.cpp:
(JSC::DFG::Graph::tryEnsureAbsence):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):

Canonical link: https://commits.webkit.org/301376@main
